### ISSUE TITLE
Wrapping standalone.xml

### ISSUE THREAD

**Issue by hwellmann** [wildfly-swarm/wildfly-swarm#4]
Given a web app and a matching `standalone.xml` configuration for vanilla WildFly containing some datasources and JMS queues etc. required by the application, why not simply package the working and tested  `standalone.xml` into the fat JAR?

I don't see the point in having to mentally translate (parts of) the XML into some kind of Java Container Configuration DSL.

**Comment by bobmcwhirter** [comment]
Sure! I think that's a reasonable request for certain. 

--

**Comment by heiko-braun** [comment]
The WF processing chain for XML conceptually works like this: 

XML > Parser > DMR > Bootstrap

For Swarm it's currently:

Fraction > Java Config >  DMR > Bootstrap

We would have to see how the XML file fit's into the picture, because it has no clear mapping to the fractions. Furthermore the XML can contain more subsystems configuration that does not correspond to a fraction.

Food for thought: if a fraction could depict which parts of the xml are relevant to it, the it could process it like:

XML > Fraction > Parser > DMR > Bootstrap (basically bypassing the Java config step), whereas XML > Fraction would a 1:n relationship, meaning that multiple fractions will process a single xml config and just consume those parts that are relevant to it.

**Comment by heiko-braun** [comment]
I'll take a stab at this.

**Comment by heiko-braun** [comment]
Some background info: https://docs.jboss.org/author/display/WFLY10/Parsing+and+marshalling+of+the+subsystem+xml

**Comment by heiko-braun** [comment]
Two things are needed:

a) Mapping of WF extensions swarm fractions: to  https://github.com/wildfly/wildfly-core/blob/master/subsystem-test/framework/src/main/java/org/jboss/as/subsystem/test/SubsystemTestDelegate.java#L168

b) The actual parsing routin that creates the ModelNode "add" operations: https://github.com/wildfly/wildfly-core/blob/master/subsystem-test/framework/src/main/java/org/jboss/as/subsystem/test/SubsystemTestDelegate.java#L216

**Comment by bobmcwhirter** [comment]
Could we somehow use StandaloneXmlParser from WF and skip the whole
fractions bits? If you provide an .xml we ignore all default fractions and
just use your xml? Or did we want to include all the other fraction stuff
just not the ModelNodes they produce?

--
Sent from my iPhone


On Nov 6, 2015, at 6:08 AM, Heiko Braun <notifications@github.com> wrote:

Two things are needed:

a) Mapping of WF extensions swarm fractions: to
https://github.com/wildfly/wildfly-core/blob/master/subsystem-test/framework/src/main/java/org/jboss/as/subsystem/test/SubsystemTestDelegate.java#L168

b) The actual parsing routin that creates the ModelNode "add" operations:
https://github.com/wildfly/wildfly-core/blob/master/subsystem-test/framework/src/main/java/org/jboss/as/subsystem/test/SubsystemTestDelegate.java#L216

—
Reply to this email directly or view it on GitHub
<https://github.com/wildfly-swarm/wildfly-swarm/issues/4#issuecomment-154382612>
.

**Comment by heiko-braun** [comment]
At first glance it seems that fractions provide more then just configuration. I.e. service activators, default deployments, etc. Then there is also the fractions that don't map to wildfly subsystems. 

The first question would be if users still need to declare fractions in the main method. That would allow us to determine what parts of the container are needed. These fractions would then retrieve there config from the XML instead of the Java API. 

An alternative approach would be to not declare fractions in main(), but  use the sheer existance of XML elements to determine which fractions should be installed. 

The former approach is more congruent with the way swarm currently works and just focuses on the configuration of a fraction through XML. My gut feeling tells me this is the right apporach.

**Comment by heiko-braun** [comment]
An example for approach a) in presudo code:

A user declares three fractions in main():
```
main {
 jpa();
 datasources();
 web();
}
```

However the XML contains much more subsystems:

```
<standalone-server>
 <messaging/>
 <Jpa/>
 <datasources/>
 <web/>
 <infinispan/>
</standalone-server>
```

Relying on this configuration in approach a) would just install jpa, datasources and web fractions and pick those parts out of the XML that matter to these fractions. The other subsystems (infinispan and messaging) that are declared in the XML would then be simply ignored.

**Comment by bobmcwhirter** [comment]
So, one thing we could do...

Our default main would still add all default fractions based upon classpath/dependencies.

All the service-activators and such (like the jolokia.war deployment) would then still occur.

We'd consume standalone.xml using WF's StandaloneXmlParser for the list of ModelNodes to configure (subverting the fraction#getList() we normally use).

Possibly each fraction could produce enough metadata so we could detect if it was configured via standalone.xml, and if not, then use getList() to do the marshalling stuff against the default fractions (which roughly match default standalone.xml)

Adding something like this, which could return `null` for non-subsystem fractions

    fractionConfiguration.getSubsystemName()

could be sufficient for us to detect if standalone.xml provided the config for a given fraction.

If the user provides a `standalone.xml` instead of `main()`, he would still need to provide the right dependencies in his pom.xml to ensure all the right jars are on the classpath or bundled into the uberjar.

I'll stop rambling now.

**Comment by heiko-braun** [comment]
>> default fractions based upon classpath/dependencies

right, that's something I didn't consider. So we agree that it all begins with declaration of a fraction either implicitly through maven dep's or explicitly through container.fraction() in main().

**Comment by kenfinnigan** [comment]
I might be mistaken, but most of this discussion appears to be around runtime use of standalone.xml.

Any thoughts on using standalone.xml at build time to actually determine the fractions to be included in uber jar?

Might be nice for the developer to not even need to concern themselves with which fractions they need to include to make standalone.xml "work".

If we're providing a standalone.xml option, would a main() even be supported for that case? Seems to be mixing worlds and might be confusing.

**Comment by bobmcwhirter** [comment]
Yes!

I think ultimately it's `standalone.xml` XOR `main()`.

If you don't have a `main()` then you aren't use the various fraction APIs directly from user code, so we don't need them explicitly mentioned in the user's `pom.xml`.

We still need to know them for `wildfly-swarm:package`, but if we had a way to determine the dependency list from scanning the `standalone.xml` at build time, then we could magically assemble it from super simple POMs.

That being said, the user's app may still need JAX-RS APIs, which currently they get transitively from `wildfly-swarm-jaxrs`, but they could go back to the traditional method of including the JAX-RS APIs as `<scope>provided</scope>` in their own POM.

The downside to this is attempting to run `org.wildfly.swarm.Swarm` default mainClass from their IDE.  This would fail without the various `wildlfy-swarm-*` jars on their classpath.  

@tobias had asked about `swarmtool` auto-scanning type of functionality in the `wildfly-swarm:package` mojo the other day also.

So I think this is something to spec out or prototype, probably in this order:

1. Support a `standalone.xml` with the assumption the user is still specifying all `wildfly-swarm-*` dependencies, purely for run-time configuration of `ModelNode` stuff.  Still probably some alteration to `wildfly-swarm:package` to replace `<mainClass>` with `<xmlConfig>` and the resulting jar uses a smarter default `main()` to handle that case.

    <configuration>
        <xmlConfig>path/to/my/standalone.xml</xmlConfig>
    </configuration>

2. Then start using the `<xmlConfig>` possibly at build-time to keep the user from having to specify the `wildfly-swarm-*` dependencies

But, as always, whoever does the work gets to make the decisions.

**Comment by heiko-braun** [comment]
@kenfinnigan raises a valid point, however I think that the rational behind @hwellmann inital request was to replace the Java config 'DSL' with XML configuration. That's one use case and the primary one for me at this point. 

There are other, related use cases as pointed out by @kenfinnigan and @bobmcwhirter, but I think we should not throw them into the same basket. From my point of view standalone.xml is an in appropriate vehicle to determine the functional scope needed to run an application. 

It may be possible to somehow "magically" determine the fractions and maven dependencies that ultimately describe the bits that make up the server runtime, but I think that conceptually fractions serve this purpose in a better way.  The XML file lacks comprehensible boundaries for a user to determine what goes into the server runtime. Fractions and their layout as maven dependencies  provide a clear and easy to grasp separation of concerns.

**Comment by heiko-braun** [comment]
Re-reading @bobmcwhirter last comment i think we should break it up into several issues. As a first step I would suggest we strive for a solution that allows XML to replace the java config-api, but anything else (maven deps, fractions, packaging) remains the same.

**Comment by kenfinnigan** [comment]
I'm not sure I'm grokking the value of supporting standalone.xml for config only. It still means the developer needs to understand fractions and what dependencies to include in their project.

Maybe it's a required stepping stone to the other pieces, but I'm not sure of its benefit to an end user.

**Comment by heiko-braun** [comment]
That's an interesting discussion ;) 

To me fractions and maven deps are the constituent parts of swarm and I would have never thought of them as a hindrance. But I see your point.

**Comment by hwellmann** [comment]
I agree with @kenfinnigan's last comment.

The basic question really is, what do I gain from using WildFly Swarm at all? See questions in #44 by @agoncal and myself. I've been watching this project for a couple of months now and I still don't have an answer for that. Especially in comparison to Docker containers.

And if there *is* a benefit, what is the least effort path from a working application with a vanilla WildFly installation and a `standalone.xml` or `domain.xml` configuration to a Swarm fat jar?

**Comment by heiko-braun** [comment]
@hwellmann 

> Especially in comparison to Docker containers.

You mean in comparison to stock wildfly in a docker container?

**Comment by hwellmann** [comment]
Yes, exactly.

**Comment by kenfinnigan** [comment]
As part of a migration to JIRA for issue tracking, this issue has been moved to https://issues.jboss.org/browse/SWARM-14
